luci-app-travelmate: support new workflow

dibdot opened the pull request:
* Still work in progress for #1769 - do not apply!

Signed-off-by: Dirk Brenken <dev@brenken.org>

dibdot commented:
@jow- I've removed the simple form stuff and tried to use the new worklfow ... it works - without reload at the end.

jow- commented:
@dibdot - ah great, that even better that what I've came up with:

```diff
diff --git a/applications/luci-app-travelmate/luasrc/model/cbi/travelmate/overview_tab.lua b/applications/luci-app-travelmate/luasrc/model/cbi/travelmate/overview_tab.lua
index 7b6b9658c..e2f4e3f00 100644
--- a/applications/luci-app-travelmate/luasrc/model/cbi/travelmate/overview_tab.lua
+++ b/applications/luci-app-travelmate/luasrc/model/cbi/travelmate/overview_tab.lua
@@ -19,8 +19,12 @@ m = Map("travelmate", translate("Travelmate"),
        .. "<a href=\"%s\" target=\"_blank\">"
        .. "see online documentation</a>", "https://github.com/openwrt/packages/blob/master/net/travelmate/files/README.md"))
 
+-- We might modify these configs too:
+m:chain("network")
+m:chain("firewall")
+
 function m.on_after_commit(self)
-       luci.sys.call("env -i /etc/init.d/travelmate restart >/dev/null 2>&1")
+       uci:apply(true)
        luci.http.redirect(luci.dispatcher.build_url("admin", "services", "travelmate"))
 end
 
@@ -42,25 +46,24 @@ if uplink == "" then
        btn.disabled = false
 
        function btn.write(self, section)
+               -- Do uci changes. Note that we must not commit here as the rollback handler
+               -- needs to see "unsaved changes" in order to roll them back.
                local iface = o:formvalue(section)
                if iface then
                        uci:set("travelmate", section, "trm_iface", iface)
-                       uci:save("travelmate")
-                       uci:commit("travelmate")
                        local net = nw:add_network(iface, { proto = "dhcp" })
                        if net then
-                               nw:save("network")
-                               nw:commit("network")
                                local zone = fw:get_zone_by_network("wan")
                                if zone then
                                        zone:add_network(iface)
-                                       fw:save("firewall")
-                                       fw:commit("firewall")
                                end
                        end
-                       luci.sys.call("env -i /bin/ubus call network reload >/dev/null 2>&1")
                end
-               luci.http.redirect(luci.dispatcher.build_url("admin", "services", "travelmate"))
+
+               -- Since we're prematurely terminating the cbi logic flow here and since we trigger
+               -- our modifications in a button write callback and not via cbi.apply, we do need to
+               -- tell cbi to trigger apply/rollback.
+               m.flow.autoapply = true
        end
        return m
 end
```

jow- commented:
See also b0f394657003d4dd954e316152498e23ad2bd4ba

dibdot commented on applications/luci-app-travelmate/luasrc/model/cbi/travelmate/overview_tab.lua:
```diff
@@ -18,49 +18,37 @@ m = Map("travelmate", translate("Travelmate"),
+					zone:add_network(value)
 				end
 			end
 			luci.sys.call("env -i /bin/ubus call network reload >/dev/null 2>&1")
```
the only caveat is the not working reload at the end to get the "normal" map view after interface creation. 
One additional question regarding your latest change, "hidebackbtn" is a controller property?

dibdot replied:
"hidebackbtn" works as expected - thanks!

jow- replied:
I think we can improve even further, on top of this PR, please try
```diff
diff --git a/applications/luci-app-travelmate/luasrc/model/cbi/travelmate/overview_tab.lua b/applications/luci-app-travelmate/luasrc/model/cbi/travelmate/overview_tab.lua
index 734f83e1c..39a461f8f 100644
--- a/applications/luci-app-travelmate/luasrc/model/cbi/travelmate/overview_tab.lua
+++ b/applications/luci-app-travelmate/luasrc/model/cbi/travelmate/overview_tab.lua
@@ -18,7 +18,11 @@ m = Map("travelmate", translate("Travelmate"),
        .. translatef("For further information "
        .. "<a href=\"%s\" target=\"_blank\">"
        .. "see online documentation</a>", "https://github.com/openwrt/packages/blob/master/net/travelmate/files/README.md"))
-m.redirect = luci.dispatcher.build_url("admin", "services", "travelmate")
+
+m.apply_on_parse = true
+m.on_apply = function()
+       luci.http.redirect(luci.dispatcher.build_url("admin", "services", "travelmate"))
+end
 
 -- Interface Wizard
 
```

in conjunction with 9693585ac911ae4cb4067dc506fac8132abdeb73 (don't forget to `rm -r /tmp/luci-*`)

dibdot commented:
I've updated my PR as requested and added your latest changes as well - but no success. The http.redirect bumps in to early and the changes gets reverted after the apply timeout (I don't see any timer of the new workflow at all).

jow- commented:
@dibdot - I added some changes and timing fixes to the apply patchset (also squashed all apply/rollback and cbi related things into one patch for easier testing) and with that it seems to work okay for me.

dibdot commented:
@jow- tested successfully with latest patchset. To get in the cbi/apply workflow I have to raise the timeout to '5' (travel router GL-AR750), please consider this is as more "save" default value.

This PR is now ready to merge (after #1769)

dibdot commented:
@jow- Thanks for your support!

jow- commented:
@dibdot - thanks for retesting. Just be sure, you do mean `option timeout` (the timeout value for XHR calls)? Or do you mean `option holdoff` (the amount of seconds to wait until beginning the confirmation attempts)?

dibdot commented:
yep, I have to change the option timeout from 3 to 5, for the remaining options I used the defaults.
